kafka-connect-s3/src/main/java/io/confluent/connect/s3/storage/S3OutputStream.java

```diff
@@ -175,6 +186,60 @@ private MultipartUpload newMultipartUpload() throws IOException {
     }
   }
 
+  protected static class RetryWithExponentialBackoff {
+    private int count;
+    private int retries = 1;
+    private Throwable cause;
+    private String errorMsg;
+
+
+    private RetryWithExponentialBackoff(int count, String errorMsg){
+      this.count = count;
+      this.errorMsg = errorMsg;
+    }
+
+     /**
+      *
+      * @param runnable The method to run with retries
+      * @param count How many times to retry
+      * @param errorMsg Error message to show
+      * @return Returns either this instance or throws ConnectException if it failed more then given count
+      */
+    public static RetryWithExponentialBackoff blocking(Runnable runnable, int count, String errorMsg){
+      RetryWithExponentialBackoff retry = new RetryWithExponentialBackoff(count, errorMsg);
+      retry.execute(runnable);
+      return retry;
+    }
+
+    private RetryWithExponentialBackoff execute(Runnable runnable){
+      if(retries > count){
+        throw new ConnectException(String.format("Giving up after failing %d times", count), cause);
+      }
+      try {
+        runnable.run();
+      } catch(SdkClientException e){
+          cause = e;
+          log.error(errorMsg + ", attempt: "+retries, cause);
+          retries++;
+          try {
+            Thread.sleep(100*2^count);
+          } catch (InterruptedException e1) {
+            log.error("Interrupted while sleeping due to retry", e1);
+          }
+          execute(runnable);
```

Comment:
This should really just be a loop, not a recursive call as using up the stack is unnecessary for this.